lua/SS13.lua

```diff
@@ -65,12 +65,25 @@ function SS13.register_signal(datum, signal, func, make_easy_clear_function)
 	local path = {"SS13", "signal_handlers", ref, signal, callback_ref, "func"}
 	callback:set_var("arguments", {path})
 	if not SS13.signal_handlers[ref]["_cleanup"] then
-		local cleanup_path = {"SS13", "signal_handlers", ref, "_cleanup"}
+		local cleanup_path = {"SS13", "signal_handlers", ref, "_cleanup", "func"}
 		local cleanup_callback = SS13.new("/datum/callback", SS13.state, "call_function_return_first", cleanup_path)
 		cleanup_callback:call_proc("RegisterSignal", datum, "parent_qdeleting", "Invoke")
-		SS13.signal_handlers[ref]["_cleanup"] = function(datum)
+		SS13.signal_handlers[ref]["_cleanup"] = {func=function(datum)
 			SS13.signal_handler_cleanup(datum)
 			dm.global_proc("qdel", cleanup_callback)
+		end, callback=cleanup_callback}
+	end
+	if signal == "parent_qdeleting" then --We want to make sure that the cleanup function is the very last signal handler called.
+		local comp_lookup = datum:get_var("comp_lookup")
+		if comp_lookup then
+			local lookup_table = comp_lookup:to_table()
+			local lookup_for_signal = lookup_table.parent_qdeleting
+			if pcall(function() return lookup_for_signal.len end) then
```

Comment:
Why the pcall? if its because `lookup_for_signal` can be nil (but not worse), then you should instead `if lookup_for_signal ~= nil and lookup_for_signal.len then`

...but now that doesn't make sense, because that seems like it'll always pass if lookup_for_signal isn't nil. 0 is truthy in Lua.

Reply:
`lookup_for_signal` can be a datum, because `/datum/proc/RegisterSignal` will set `comp_lookup[sigtype]` to a reference to the listener if there are no listeners for that signal. `comp_lookup[sigtype]` will only become a list after more than one listener has been registered to a signal. With that said, I just realized that instead of a pcall, I can do `SS13.istype(lookup_for_signal, "/datum")`

Reply:
Yeah you should do that